torch/nn/modules/activation.py

```diff
@@ -1034,6 +1056,66 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
             `batch_first` argument is ignored for unbatched inputs.
         """
         is_batched = query.dim() == 3
+        why_not_fast_path = ''
+        if not is_batched:
+            why_not_fast_path = f"input not batched; expected query.dim() of 3 but got {query.dim()}"
+        elif query is not key or key is not value:
+            why_not_fast_path = "non-self attention was used (query, key, and value are not the same Tensor)"
+        elif self.training:
+            why_not_fast_path = "training is enabled"
+        elif not self.batch_first:
+            why_not_fast_path = "batch_first was not True"
+        elif self.bias_k is not None:
+            why_not_fast_path = "self.bias_k was not None"
+        elif self.bias_v is not None:
+            why_not_fast_path = "self.bias_v was not None"
+        elif self.dropout:
+            why_not_fast_path = "dropout was not zero"
+        elif self.add_zero_attn:
+            why_not_fast_path = "add_zero_attn was enabled"
+        elif not self._qkv_same_embed_dim:
+            why_not_fast_path = "_qkv_same_embed_dim was not True"
+        elif query.is_nested and (key_padding_mask is not None or attn_mask is not None):
+            why_not_fast_path = "key_padding_mask and attn_mask are not supported with NestedTensor input"
+        elif not query.is_nested and key_padding_mask is not None and attn_mask is not None:
+            why_not_fast_path = "key_padding_mask and attn_mask were both supplied"
+
+        if not why_not_fast_path:
+            tensor_args = (
+                query,
+                key,
+                value,
+                self.in_proj_weight,
+                self.in_proj_bias,
+                self.out_proj.weight,
+                self.out_proj.bias,
+            )
+            # We have to use list comprehensions below because TorchScript does not support
+            # generator expressions.
+            if torch.overrides.has_torch_function(tensor_args):
+                why_not_fast_path = "some Tensor argument has_torch_function"
```

Comment:
why would it be supported? I cannot guarantee at all that torch_functions will be called.

Reply:
The codegen that binds your code to python is adding the torch_function handler for you :)

Reply:
I am thinking of torch functions for *other* operators. torch_function is essentially COME FROM. Somebody could use it to get a callback inside MHA, and the fast path will not call those callbacks, which is a behavior change, so we can't use the fast path.

Reply:
Put another way, if somebody was making MHA work by `__torch_function__`ing the pieces it calls into previously, forcing them to execute the fast path will break their code.

Reply:
you can't torch_function all of MHA. it makes no sense to torch_function just the fast path. If you enter the fast path using some kind of weird tensor-like that has torch_functions, it is going to crash. Therefore, we cannot enter the fast path in that case at all; if we tried to, it would break code that previously worked.